src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,248 @@
+  private static final long   MINIMUM_CHUNK_SIZE        = 2000L;
+  private static final double JITTER_MAX                = 0.20;
+
+  private final AccountsManager               accountsManager;
+  private final DirectoryReconciliationClient reconciliationClient;
+  private final DirectoryReconciliationCache  reconciliationCache;
+  private final String                        workerId;
+  private final SecureRandom                  random;
+
+  private ScheduledExecutorService executor;
+
+  public DirectoryReconciler(DirectoryReconciliationClient reconciliationClient,
+                             DirectoryReconciliationCache reconciliationCache,
+                             AccountsManager accountsManager) {
+    this.accountsManager      = accountsManager;
+    this.reconciliationClient = reconciliationClient;
+    this.reconciliationCache  = reconciliationCache;
+    this.random               = new SecureRandom();
+    this.workerId             = generateWorkerId(random);
+  }
+
+  private static String generateWorkerId(SecureRandom random) {
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    return Hex.toString(workerIdBytes);
+  }
+
+  @Override
+  public void start() {
+    start(Executors.newSingleThreadScheduledExecutor());
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executor) {
+    this.executor = executor;
+
+    try {
+      reconciliationCache.cleanUpWorkerSet(WORKER_TTL_MS);
+    } catch (Exception ex) {
+      logger.warn("failed to clean up worker set", ex);
+    }
+
+    this.executor.scheduleWithFixedDelay(new PingTask(workerId, reconciliationCache), 0, WORKER_TTL_MS / 2, TimeUnit.MILLISECONDS);
+    scheduleWithJitter(CHUNK_INTERVAL);
+  }
+
+  @Override
+  public void stop() {
+    this.executor.shutdown();
+
+    try {
+      this.executor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ex) {
+      throw new AssertionError(ex);
+    }
+
+    try {
+      reconciliationCache.leaveWorkerSet(workerId);
+    } catch (Exception ex) {
+      logger.warn("failed to remove from worker set", ex);
+    }
+  }
+
+  @Override
+  public void run() {
+    boolean accelerate  = false;
+    long    workerCount = 1;
+
+    try {
+      if (reconciliationCache.lockActiveWorker(workerId, CHUNK_INTERVAL)) {
+        try {
+          if (processChunk()) {
+            accelerate = reconciliationCache.isAccelerated();
+          }
+        } finally {
+          reconciliationCache.unlockActiveWorker(workerId);
+        }
+      }
+
+      if (!accelerate) {
+        workerCount = Math.max(reconciliationCache.getWorkerCount(WORKER_TTL_MS), 1);
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    if (!accelerate) {
+      scheduleWithJitter(CHUNK_INTERVAL * workerCount);
+    } else {
+      scheduleWithJitter(ACCELERATE_CHUNK_INTERVAL);
+    }
```

Comment:
Keeping track of the number of workers seems like.... a lot of work. Would it be simpler for the active worker to  set the time the next chunk of work should occur (or the time at which the last chunk of work did occur) and have all workers check that once a second or whatever?

Reply:
yeah now that i have the "active worker" lock, that would make sense: just have every worker wake up (with jitter) every chunk interval to try to do work (checking the last work time like you mentioned). though instead of checking once a second, if there's an active worker, i think i will just have workers wait for another chunk interval (with jitter)

Reply:
see e40a7968ae83. i love deleting code